src/librustc/infer/error_reporting/mod.rs

```diff
@@ -966,3 +712,515 @@ impl<'tcx> ObligationCause<'tcx> {
+                                for item_id in &mod_.item_ids {
+                                    if let hir::map::NodeItem(item) = self.tcx.hir.get(item_id.id) {
+                                        // We only want to add items that have been exported to the
+                                        // type alias table, check visibility.
+                                        match item.vis {
+                                            hir::Visibility::Public |
+                                            hir::Visibility::Crate => {
+                                                let ty = self.tcx.node_path_str(item.id);
+
+                                                self.full_imports.insert(format!("{}", ty),
+                                                                         format!("{}", item.name));
+                                            }
+                                            _ => (),  // item not visible, don't alias
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    } else {
+                        // Importing from a different crate.
+                        for child in self.tcx.cstore().item_children_untracked(def_id,
+                                                                               self.tcx.sess) {
+                            handle_external_def(self.tcx,
+                                                &mut external_mods,
+                                                child.def,
+                                                &mut self.full_imports)
+                        }
+                    }
+                }
+                if let Err = path.def {  // external crate
+                    // Cross-crate:
+                    for cnum in self.tcx.cstore().crates_untracked() {
+                        let crate_name = self.tcx.cstore().crate_name_untracked(cnum);
+                        if Some(crate_name) == path.segments.first().map(|s| s.name) {
+                            let def_id = DefId {
+                                krate: cnum,
+                                index: hir::def_id::CRATE_DEF_INDEX,
+                            };
+                            debug!("visit_item crate cnum {:?} {}", cnum, crate_name);
+                            // Filter out crates not imported in this context.
+                            handle_external_def(self.tcx,
+                                                &mut external_mods,
+                                                Mod(def_id),
+                                                &mut self.full_imports);
+                        }
+                    }
+                }
+            }
+            _ => (), // Ignore non-import items
+        }
+    }
+}
+
+/// This struct knows how to print out a single type for human consumption, compare two types and
```

Comment:
Can this replace ppaux?

Reply:
That is my eventual objective. I have moved the type argument elision code away from `TyCtxt` in order to isolate as much from it as possible in order to unify it with the code in `ppaux`. Do you think I should strive to do it in this PR?